### PR TITLE
Fast path binary ops in fake tensor

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94047

Fast path execution of a few binary ops in fake tensor, to speed up trace time. When testing `python benchmarks/dynamo/timm_models.py --accuracy --timing --backend aot_eager --dynamic-shapes --float32 --only hrnet_w18`, I get the following trace speedup.

Before:

```
cuda eval  hrnet_w18                           PASS
TIMING: entire_frame_compile:53.97591 backend_compile:33.60832
STATS: call_* op count: 1369 | FakeTensor.__torch_dispatch__:4995 | FakeTensorMode.__torch_dispatch__:89985 | ProxyTorchDispatchMode.__torch_dispatch__:3010
```

After:

```
cuda eval  hrnet_w18                           PASS
TIMING: entire_frame_compile:40.18931 backend_compile:25.28828
STATS: call_* op count: 1369 | FakeTensor.__torch_dispatch__:4995 | FakeTensorMode.__torch_dispatch__:69478 | attempt fast:4399 | fast is_contiguous:4399 | ProxyTorchDispatchMode.__torch_dispatch__:3010
```

My experiment notebook can be found at https://docs.google.com/document/d/1_dTIQUwjIVnEWmiFAavJQYVF8uzXqD9Dk6b9gGQLF_U/edit#

This is not the "most" optimized version of the code; compared with Horace/Voz roofline experiment:

```
diff --git a/torch/_subclasses/fake_tensor.py b/torch/_subclasses/fake_tensor.py
index e3bf545f3b8..395942c6ffe 100644
--- a/torch/_subclasses/fake_tensor.py
+++ b/torch/_subclasses/fake_tensor.py
@@ -774,6 +774,10 @@ class FakeTensorMode(TorchDispatchMode):
     def __torch_dispatch__(self, func, types, args=(), kwargs=None):
         kwargs = kwargs if kwargs else {}
 
+        with no_dispatch():
+            if func in {aten.mul.Tensor, aten.add.Tensor, aten.sub.Tensor, aten.relu.default}:
+                return FakeTensor(self, torch.empty(args[0].shape, device='meta'), device='cuda')
+
         if func == torch.ops.prim.device.default:
             assert len(args) == 1 and isinstance(args[0], FakeTensor)
             if args[0].fake_mode.in_kernel_invocation:
```

I am still leaving about 5s of trace time improvement on the table (3s of which is attributable to not yet handling relu.)

The implementation here is based off of https://github.com/pytorch/pytorch/pull/93118/ but I modeled the short circuit logic off of TensorIterator's implementation, for ease of code review and correctness verification. However, there are some important divergences:

* Traditional fast setup in TensorIterator only short circuits if the shapes of all input elements are equal. On hrnet_w18, only 5% of fastpath'ed binary operators actually satisfy this. So instead, I compute the broadcasted shape, but then I only allow the fast path if (1) at least one input tensor has a shape that is exactly the output size, and (2) all the tensors are contiguous (or if all the tensors are channels last).
* I had to manually adjust the logic to handle wrapped numbers (which ordinarily are handled by wrapping into tensors). I think I got this right.

Some evidence that this heuristic is correct is here in: https://gist.github.com/ezyang/b22fa7b72b7349137211d8dc7041f758 I exhaustively test all dim=3 tensors with sizes [1, 2] and show that we get the same significant strides between PrimTorch and the new algorithm. In fact, there ARE differences between this algorithm and PrimTorch, but in fact this algorithm agrees with TensorIterator where PrimTorch is wrong (sample case: size=(1, 1, 2), stride=(1, 1, 1), stride=(1, 1, 1))

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

**Review by albanD — COMMENTED** [review]
Could you give more details on the kind of gains we get from this? This does add significant complexity.

**Comment by ezyang** [comment]
PR description updated!

File: torch/utils/_stats.py [hunk 1]
```diff
@@ -3,8 +3,13 @@
 # AND SCRUB AWAY TORCH NOTIONS THERE.
 import collections
 import functools
+from typing import OrderedDict
 
-simple_call_counter = collections.OrderedDict()
+simple_call_counter: OrderedDict[str, int] = collections.OrderedDict()
```
Comment (eellison) [inline torch/utils/_stats.py:8]: why do we need OrderedDict ? isn't dict already ordered ?

File: torch/_subclasses/fake_tensor.py [hunk 2]
```diff
@@ -509,6 +530,168 @@ def convert(t, mem_fmt):
...
+            count_label(f"slow {msg}")
+            with mode:
+                return slow_ref(*args, **kwargs)
+
+        count_label("attempt fast")
+        # == Fast path (based off of TensorIterator fast path) ==
+        operands = args
+
+        # compute_shape
+        has_scalars = False
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:575]: it's a lot of code duplication of logic that exists elsewhere :/ but if it's much slower to reuse code, i guess that's life. In any case, could we test that ?

**Review by eellison — COMMENTED** [review]
have you tested short circuiting but reusing more of the methods from prims / elsewhere for this ? Would be nice to cut down on some of the duplication. I think a lot of the speedup would still be applicable

File: torch/_subclasses/fake_tensor.py [hunk 3]
```diff
@@ -766,6 +949,13 @@ def __init__(
         self.shape_env = shape_env
 
     def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+        try:
+            return self.dispatch(func, types, args, kwargs)
+        except TypeError:
+            log.exception("fake tensor raised TypeError")
+            raise
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:956]: When I was working on this PR I sometimes messed up my short circuit logic and triggered a TypeError. This TypeError was silently swallowed. Now I get a log for it.

File: torch/_subclasses/fake_tensor.py [hunk 4]
```diff
@@ -509,6 +530,168 @@ def convert(t, mem_fmt):
...
+                continue
+            is_contiguous = is_contiguous and op.is_contiguous(
+                memory_format=torch.contiguous_format
+            )
+            is_channels_last = is_channels_last and op.is_contiguous(
+                memory_format=torch.channels_last
+            )
+        if is_contiguous:
+            # do contiguous
+            count_label("fast is_contiguous")
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:659]: TBH we should do more telemetry. I'm happy to remove this but this was also very useful for understand perf characteristics here.

File: torch/_subclasses/fake_tensor.py [hunk 5]
```diff
@@ -509,6 +530,168 @@ def convert(t, mem_fmt):
...
+    dimsB = len(b)
+    ndim = max(dimsA, dimsB)
+    expandedSizes = [0] * ndim
+    for i in range(ndim - 1, -1, -1):
+        offset = ndim - 1 - i
+        dimA = dimsA - 1 - offset
+        dimB = dimsB - 1 - offset
+        sizeA = a[dimA] if dimA >= 0 else 1
+        sizeB = b[dimB] if dimB >= 0 else 1
+        assert sizeA == sizeB or sizeA == 1 or sizeB == 1
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:558]: oh oops, yeah, I will change this

File: torch/utils/_stats.py (hunk 1 above)
Comment (ezyang) [reply torch/utils/_stats.py:8]: ask @voznesenskym

File: torch/_subclasses/fake_tensor.py (hunk 2 above)
Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:575]: I'm working on the testing right now. It's slower to reuse code: PrimTorch doesn't actually implement a fastpath, and this is different logic than TensorIterator.

**Comment by ezyang** [comment]
Now updated with some evidence the heuristic is OK; see bottom of PR desc=

**Comment by ezyang** [comment]
Here are the most improved models with this change:

image

and the least improved:

image

eca_halonext26ts is also interesting: a long running model that isn't helped much

image

File: torch/utils/_stats.py [hunk 6]
```diff
@@ -3,8 +3,13 @@
...
+
+def count_label(label):
+    prev = simple_call_counter.setdefault(label, 0)
+    simple_call_counter[label] = prev + 1
```
Comment (albanD) [inline torch/utils/_stats.py:12]: You could use a defaultdict and just += 1 here.

File: torch/_subclasses/fake_tensor.py [hunk 7]
```diff
@@ -31,6 +39,18 @@
 
 CONSTANT_NUMEL_LIMIT = 1
 
+CNT = 0
+
+
+class Increment:
```
Comment (albanD) [inline torch/_subclasses/fake_tensor.py:45]: That looks shady. Is it just for debugging?

**Review by albanD — COMMENTED** [review]
Debug code needs to go but sounds ok otherwise.

File: torch/_subclasses/fake_tensor.py (hunk 7 above)
Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:45]: It's for the debug logging. I can delete the debug logging code but, IDK, I thought it was pretty useful, I feel like we should at least let it hang out in master a bit.

File: torch/utils/_stats.py (hunk 6 above)
Comment (ezyang) [reply torch/utils/_stats.py:12]: IDK why @voznesenskym didn't make this a defaultdict, I was minimizing changes here

File: test/functorch/test_aotdispatch.py [hunk 8]
```diff
@@ -2111,6 +2111,7 @@ def forward(self, x, y):
 
         self.assertExpectedInline(shape_env.format_guards(), """\
  - Eq(s1, 20)
+ - Ne(s0, 30)
```
[the remark this answers is not in the document]
Comment (ezyang) [reply test/functorch/test_aotdispatch.py:2114]: No, it's not really OK. Need to figure out where it came from.

File: torch/_subclasses/fake_tensor.py (hunk 7 above)
Comment (albanD) [reply torch/_subclasses/fake_tensor.py:45]: I don't think this kind of logging is useful in general tbh. it just ends up bloating the code and generate so much lines that it is impossible to find anything in there.

But Dynamo and Inductor are already going down that path so I won't actively block this.

**Comment by albanD** [comment]
btw the debug code I said need to go is the one raised by Elias above that you said you'll remove before merging.

**Comment by ezyang** [comment]
OK, to be clear, @eellison do you want it removed? I would prefer it to stay but if someone says "please remove" I will remove.

**Comment by albanD** [comment]
I think the worse error message is worth fixing.
The others are all metrics collection and logging. So fine.

File: torch/_subclasses/fake_tensor.py [hunk 9]
```diff
@@ -509,6 +529,179 @@ def convert(t, mem_fmt):
...
+        dimA = dimsA - 1 - offset
+        dimB = dimsB - 1 - offset
+        sizeA = a[dimA] if dimA >= 0 else 1
+        sizeB = b[dimB] if dimB >= 0 else 1
+        assert sizeA == sizeB or sizeA == 1 or sizeB == 1
+        expandedSizes[i] = sizeB if sizeA == 1 else sizeA
+    return tuple(expandedSizes)
+
+
+def make_fast_binary_impl(slow_ref):
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:562]: would it be worth moving this to fake_utils ? idk

File: torch/_subclasses/fake_tensor.py (hunk 7 above)
Comment (eellison) [reply torch/_subclasses/fake_tensor.py:45]: Increment is a bit misleading. It's really something more like StackCount right, or RecursiveCount ?

File: torch/_subclasses/fake_tensor.py (hunk 2 above)
Comment (eellison) [reply torch/_subclasses/fake_tensor.py:574]: Where is the C++ -> python LLM when you need it 🤡 🤪

File: torch/_subclasses/fake_tensor.py [hunk 10]
```diff
@@ -509,6 +529,179 @@ def convert(t, mem_fmt):
...
+    def fast_binary_impl(mode, *args, **kwargs):
+        def slow(msg):
+            count_label(f"slow {msg}")
+            with mode:
+                return slow_ref(*args, **kwargs)
+
+        count_label("attempt fast")
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:569]: would we even need the count_labels if these were factored out into functions ? how slow are python profilers? or maybe just annoying to use

**Review by eellison — APPROVED** [review]
I'm fine with it I guess but if we get more telemetry we should make them not on by default. inductor has various loggings but I don't think any of them are on by default. 

I don't particularly care in either direction. Alban's concern about readability is true but we can cross that bridge when we get to it.

**Comment by ezyang** [comment]
> I'm fine with it I guess but if we get more telemetry we should make them not on by default. inductor has various loggings but I don't think any of them are on by default.

The telemetry goes into some counters which don't get printed by default. That's the same as how dynamo collects counter telemetry too.

File: torch/_subclasses/fake_tensor.py (hunk 9 above)
Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:562]: it only needs to be used here, so let's keep it here. If we want, we could make a separate module for "fake tensor op implementations"

File: torch/_subclasses/fake_tensor.py (hunk 10 above)
Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:569]: Yeah, it's a combination of slowness (for non-sampling profilers) and also annoyance (the function call will be lost in a sea of other function calls.)

File: torch/utils/_stats.py (hunk 1 above)
Comment (voznesenskym) [reply torch/utils/_stats.py:8]: I like preserving key order for things like this.

File: torch/utils/_stats.py (hunk 6 above)
Comment (voznesenskym) [reply torch/utils/_stats.py:12]: Sure, that's fine.

**Pull request closed without merging**